lms/services/vitalsource/model.py

```diff
@@ -9,23 +11,63 @@ class VSBookLocation:
     book_id: str
     """Id of the book"""
 
-    cfi: str
-    """Location within than book."""
+    cfi: Optional[str] = None
+    """Location within the book, specified as a CFI."""
+
+    page: Optional[str] = None
+    """
+    Location within the book, specified as a page number.
+
+    In most cases this will be a 1-based number, but some pages may use other
+    numbering systems (eg. Roman numerals, prefixes).
+
+    Page numbers are usually unique within a book, but this is not guaranteed.
+    In the event that a page number corresponds to multiple locations, this
+    VSBookLocation is taken to refer to the first such location.
+    """
 
     @property
     def document_url(self):
         """Get our internal representation of this location."""
 
-        return f"vitalsource://book/bookID/{self.book_id}/cfi/{self.cfi}"
+        url = f"vitalsource://book/bookID/{self.book_id}"
+
+        if self.cfi:
+            # FIXME: CFIs are currently not escaped. This could create parsing
+            # ambiguities when query params are added to the URL. If we change
+            # this we need to avoid breaking existing assignments.
+            url += f"/cfi/{self.cfi}"
```

Comment:
I didn't change escaping for CFIs in this PR to avoid introducing backwards compatibility issues. See also [this Slack thread](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1700495733366949) about whether we should reconsider how we store document metadata in the longer term.